modules/frontend/tracebyidtenantparsing.go

```diff
@@ -0,0 +1,169 @@
+package frontend
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/golang/protobuf/proto"
+	"github.com/grafana/dskit/tenant"
+	"github.com/grafana/dskit/user"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/boundedwaitgroup"
+	"github.com/grafana/tempo/pkg/model/trace"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/opentracing/opentracing-go"
+)
+
+func newTraceByIDTenantParser(cfg *TraceByIDConfig, logger log.Logger) Middleware {
+	return MiddlewareFunc(func(next http.RoundTripper) http.RoundTripper {
+		return parseQuery{
+			next:   next,
+			cfg:    cfg,
+			logger: logger,
+		}
+	})
+}
+
+type parseQuery struct {
+	next   http.RoundTripper
+	cfg    *TraceByIDConfig
+	logger log.Logger
+}
+
+func (p parseQuery) RoundTrip(r *http.Request) (*http.Response, error) {
```

Comment:
we will need this same functionality on all paths: tags, metrics, trace by id and search. i think we should try to implement this as a generic piece of middleware.

the middleware would define a "combiner" interface that it used to recombine the incoming http requests. wdyt?

Reply:
Ok. So every path then would provide implementation for this new combiner interface as per path's requirements.

Reply:
yes, but tbh i don't think this is possible atm on the metrics endpoints. maybe just return an error there for now since it's experimental?

Reply:
I check that search and trace by id have diff ways of processing requests and combining the responses. So another approach may be to pass this behaviour as a function to the middleware.

Reply:
```
type Combiner interface {
	Consume(input any) (count int)
	Result(statusCode int, statusMsg string) (*http.Response, error)
}
```
This is the Combiner interface I plan to have.

Consume is for merging responses of individual requests.
Result is for putting together final response based on merge result(and final statuscode, status message after all requests are processed).

Reply:
- `Consume()` should take an `*http.Response`. the middleware should just give it all responses it receives and rely on the interface's implementation to work out the details
- Does result need the status code/message? These values will be in each consumed http.response?

I'm thinking like:

```
type Combiner interface {
	Consume(*http.Response)
	Result() (*http.Response, error)
}
```

as discussed out of band, let's:

1) make this configurable (off by defafult)
2) only worry about the trace by id path right now

if we establish a good generic set of code we can add the others later.

Reply:
As consuming a response may result into an error too(like when reading response body or during unmarshalling the body), so Consume must return error, and then maybe a we call it Process.

`Process(*http.Response) error`

Reply:
I'm fine with either Process or Result. I think they are clear enough

Reply:
Added a cross tenant middleware which takes combiner as input, and added one such combiner for traces